### PR TITLE
Sparse Array

### PR DESCRIPTION
**Pull request by YuxingQiu**
Add new support to the sparse array. 

Updates include:
1. `SparseArrayLayout` that supports:
    - record valid sparse tiles inside the sparse map
    - query sparse map
    - query array-related info (e.g. size)
    - insert tiles according to particle positions 
2. `SparseArray` that supports:
    - array allocation
    - array access with multiple indexing style 
3. Corresponding unit tests

### CODE REVIEW

File: cajita/src/Cajita_SparseArray.hpp [hunk 1]
```diff
@@ -0,0 +1,541 @@
...
+{
+
+namespace Experimental
+{
+//---------------------------------------------------------------------------//
+// Indexing type tags
+//---------------------------------------------------------------------------//
+// H - Hierachical
+// T_C: indexing manner: tile ijk - cell ijk
+struct Index_H_T_C
```
Comment (streeve) [inline cajita/src/Cajita_SparseArray.hpp:26]: From yesterday, let's see if it's possible to differentiate overloads without these tags (using `Kokkos::Array` or `std::array`

File: cajita/src/Cajita_SparseArray.hpp [hunk 2]
```diff
@@ -0,0 +1,541 @@
...
+    inline uint64_t arraySizeTile() const
+    {
+        entity_type t;
+        return arraySizeTileImpl( t );
+    }
+
+    inline void clear() { _map.clear(); }
+
+    template <class ExecSpace, class PositionSliceType>
+    void register_sparse_map( PositionSliceType& poses )
```
Comment (streeve) [inline cajita/src/Cajita_SparseArray.hpp:104]: `poses` -> `positions`

File: cajita/unit_test/tstSparseArray.hpp [hunk 3]
```diff
@@ -0,0 +1,622 @@
...
+                            Kokkos::Array<int, 3> cell_ijk(
+                                { ti * cell_per_tile_dim + ci,
+                                  tj * cell_per_tile_dim + cj,
+                                  tk * cell_per_tile_dim + ck } );
+                            Kokkos::Array<int, 3> tile_ijk( { ti, tj, tk } );
+                            Kokkos::Array<int, 3> local_cell_ijk(
+                                { ci, cj, ck } );
+
+                            // access: cell ijk (- channel id)
+                            test_array.template get<0>( cell_ijk, 0 ) = ti;
```
Comment (streeve) [inline cajita/unit_test/tstSparseArray.hpp:335]: We should be able to reuse or imitate the `Cabana::SoA` cast functions to make this a little easier for the user

**Review by streeve — COMMENTED** [review]
Looking very good - I did a tiny amount of cleanup and had a couple questions

File: cajita/unit_test/tstSparseArray.hpp (hunk 3 above)
Comment (streeve) [reply cajita/unit_test/tstSparseArray.hpp:335]: Let's leave for now and plan to hide inside the parallel functions

File: cajita/src/Cajita_SparseArray.hpp [hunk 4]
```diff
@@ -0,0 +1,674 @@
...
+  \param sparse_map The reference to a pre-created sparse map.
+  \param EntityType The entity: Cell, Node, Face, or Edge.
+  \param bc_factor Factor used to increase allocation size for special entities.
+*/
+template <class DataTypes, class EntityType, class MeshType,
+          class SparseMapType>
+SparseArrayLayout<DataTypes, EntityType, MeshType, SparseMapType>
+createSparseArrayLayout( const std::shared_ptr<LocalGrid<MeshType>>& local_grid,
+                         SparseMapType& sparse_map, EntityType,
+                         const float bc_factor = 1.01f )
```
[the remark this answers is not in the document]
Comment (streeve) [reply cajita/src/Cajita_SparseArray.hpp:282]: Let's rename as `overallocation` for a general option and leave the specific allocation guess for face/edge for later

File: cajita/unit_test/tstSparseArray.hpp (hunk 3 above)
Comment (YuxingQiu) [reply cajita/unit_test/tstSparseArray.hpp:335]: Sounds great

File: cajita/src/Cajita_SparseArray.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (YuxingQiu) [reply cajita/src/Cajita_SparseArray.hpp]: Sure, sounds great. It's updated

**Comment by streeve** [comment]
retest this please.

**Review by streeve — APPROVED** [review]

**Pull request merged**